packages/sdk-codegen/src/specLinter.ts

```diff
@@ -0,0 +1,203 @@
+
+const diffToString = (diff: any) =>
+  diff ? `"${JSON.stringify(diff, null, 2).replace(/"/g, "'")}."` : ''
+
+export interface ComputeDiffInputs {
+  lMethod?: IMethod
+  rMethod?: IMethod
+}
+
+interface DiffRow {
+  /** Method name */
+  name: string
+  /** Method operation and path */
+  id: string
+  /** Method status in both specs */
+  lStatus: string
+  rStatus: string
+  /** Method type diff if any */
+  typeDiff: string
+  /** Method params diff if any */
+  paramsDiff: string
+}
+
+/**
+ * Computes the diff of two methods
+ * @param lMethod
+ * @param rMethod
+ */
+const computeDiff = ({ lMethod, rMethod }: ComputeDiffInputs) => {
+  if (!lMethod && !rMethod) throw new Error('At least one method is required.')
+
+  let result: DiffRow
+  if (lMethod) {
+    result = {
+      name: lMethod.name,
+      id: lMethod.id,
+      lStatus: lMethod.status,
+      rStatus: rMethod ? rMethod.status : '',
+      typeDiff: rMethod
+        ? diffToString(compareTypes(lMethod.type, rMethod.type))
+        : '',
+      paramsDiff: rMethod ? diffToString(compareParams(lMethod, rMethod)) : '',
+    }
+  } else if (!lMethod && rMethod) {
+    result = {
+      name: rMethod.name,
+      id: rMethod.id,
+      lStatus: '',
+      rStatus: rMethod.status,
+      typeDiff: '',
+      paramsDiff: '',
+    }
+  }
+
+  return result!
+}
+
+/**
+ * Given two specs, compares each method's params and response type, and returns
+ * a string representing representing a csv containing the difference if any
+ * @param lSpec
+ * @param rSpec
+ */
+
+export type preDiffCb = ({ lMethod, rMethod }: ComputeDiffInputs) => boolean
+
+/**
+ * Compares two specs and returns a diff object
+ * @param lSpec
+ * @param rSpec
+ * @param preDiffCb A callback for filtering methods prior to computing the diff
+ */
+export const compareSpecs = (
+  lSpec: IApiModel,
+  rSpec: IApiModel,
+  preDiffCb?: preDiffCb
+) => {
+  const lMethods = Object.values(lSpec.methods)
+  const rMethods = Object.values(rSpec.methods)
+  const rLength = rMethods.length
+  let rIndex = 0
+  const diff: DiffRow[] = []
+
+  lMethods.forEach((lMethod) => {
+    let rMethod = rMethods[rIndex]
+
+    if (preDiffCb && !preDiffCb({ lMethod, rMethod })) return
+
+    while (
+      lMethod.name !== rMethod.name &&
+      rMethod.name < lMethod.name &&
+      rIndex < rLength
+    ) {
+      /** Case when right method does not exist on the left */
+      if (preDiffCb && preDiffCb({ rMethod })) {
+        diff.push(computeDiff({ rMethod }))
+      }
+      rIndex += 1
+      rMethod = rMethods[rIndex]
+    }
+
+    if (lMethod.name === rMethod.name) {
+      /** Case when method exists on both sides */
+      diff.push(computeDiff({ lMethod, rMethod }))
+      rIndex = rIndex + 1 < rLength ? rIndex + 1 : rIndex
+    } else {
+      /** Case when left method does not exist on the right */
+      diff.push(computeDiff({ lMethod }))
```

Comment:
same question as above. Guess I'll need to debug this to understand it, but that indicates code smell to me